[FSDP2] enable CI for torch.compile(root Transformer)

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127832

This CI showcases FSDP2 works with `torch.compile` root model, since FSDP1 can do the same 

compiling root Transformer without AC: `pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_multi_group`

compiling root Transformer with AC: `pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_with_activation_checkpointing`

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

weifengpy committed:
[FSDP2] enable CI for torch.compile root Transformer

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -285,44 +285,11 @@ def test_train_parity_multi_group_eager(self):
-            {
-                "reshard_after_forward": [True],  # save CI time
                 "offload_policy": [
+                    None,
```
Is `offload_policy` allowed to be `None`? I thought it was only `OffloadPolicy` or `CPUOffloadPolicy` for now.

weifengpy replied:
you're right. I am changing to `OffloadPolicy` now (`None` magically worked but I will respect the type hint)

weifengpy committed:
Update on "[FSDP2] enable CI for torch.compile root Transformer"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -275,7 +274,8 @@ def _test_train_parity_single_group(self, lin_shapes: List[Tuple[int, int]]):
             self.assertEqual(losses[0], losses[1])
 
     @skip_if_lt_x_gpu(2)
```
we can support `reshard_after_forward=2` with `torch._dynamo.config.skip_fsdp_hooks=True`

it's a good time to merge `test_train_parity_multi_group_eager`, `_compile`, `cpu_offload_eager` into `test_train_parity_multi_group`.

awgu replied:
I recall that if we merged `cpu_offload_eager` with the normal eager test, sometimes the test becomes flaky for unknown reasons.

awgu replied:
It might be good to keep the CPU offloading one separate.

weifengpy replied:
will keep cpu offloading as a separate test

weifengpy committed:
Update on "[FSDP2] enable CI for torch.compile root Transformer"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -243,7 +243,6 @@ def world_size(self) -> int:
         return min(8, torch.cuda.device_count())
 
     @skip_if_lt_x_gpu(2)
-    @test_compiled_fsdp()
```
remove trivial use ucase and instead focusing on `torch.compile(Transformer)` w/o AC

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -579,16 +557,18 @@ def _test_train_parity_with_activation_checkpointing(
-            check_sharded_parity(
-                self, ref_model, model, prefixes_to_ignore=prefixes_to_ignore
-            )
+            if not testing_compile:
```
we have loss numeric parity with `torch.compile`. But does not have `.grad` numeric parity

awgu replied:
How is it that we have loss parity over multiple iterations but not gradient parity? 🤔 

I would expect the losses to diverge on the subsequent iteration then.

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -597,22 +577,11 @@ def world_size(self) -> int:
         return min(4, torch.cuda.device_count())
 
     @skip_if_lt_x_gpu(2)
```
clean up by merging `test_train_parity_with_shared_params_no_ac`, `_ac` into single `test_train_parity_with_shared_params`

weifengpy commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -1416,45 +1424,49 @@ def _test_fsdp_parity(
+                torch.distributed.barrier()
+                func.__globals__[original_fully_shard.__name__] = original_fully_shard
+                torch._dynamo.config.skip_fsdp_hooks = original_skip_fsdp_hooks
+                torch._inductor.config.compile_threads = original_compile_threads
```
equivalent to `TORCHINDUCTOR_COMPILE_THREADS=1`. This prevents flaky results with `FSDPTest`/ multi-process

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -536,7 +514,7 @@ def _test_train_parity_with_activation_checkpointing(
                 vocab_size=vocab_size,
                 max_seq_len=64,
-                dropout_p=0.1,
+                dropout_p=0,
```
Is this because `torch.compile` does not have same randomness as eager?
Should we just disable it for compile path then?

weifengpy committed:
Update on "[FSDP2] enable CI for torch.compile(root Transformer)"


This CI showcases FSDP2 works with `torch.compile` root model, since FSDP1 can do the same 

compiling root Transformer without AC: `pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_multi_group`

compiling root Transformer with AC: `pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_with_activation_checkpointing`

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP2] enable CI for torch.compile(root Transformer)"


This CI showcases FSDP2 works with `torch.compile` root model, since FSDP1 can do the same 

compiling root Transformer without AC: `pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_multi_group`

compiling root Transformer with AC: `pytest test/distributed/_composable/fsdp/test_fully_shard_training.py -k test_train_parity_with_activation_checkpointing`

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge

weifengpy commented:
@pytorchmergebot merge
